File: sdk/typescript/src/auth/TokenManager.ts
import axios from 'axios';
import { AuthenticationError } from '../errors';

/**
 * Manages OAuth2 client-credentials token lifecycle.
 * Caches token in memory, refreshes 60s before expiry.
 */
export class TokenManager {
  private cachedToken: string | null = null;
  private expiresAt: number = 0;
  private static readonly EXPIRY_BUFFER = 60;

  constructor(
    private readonly baseUrl: string,
    private readonly clientId: string,
    private readonly clientSecret: string
  ) {}

  async getToken(): Promise<string> {
    if (this.cachedToken && Date.now() / 1000 < this.expiresAt - TokenManager.EXPIRY_BUFFER) {
      return this.cachedToken;
    }
    return this.refresh();
  }

  async refresh(): Promise<string> {
    try {
      const response = await axios.post(
        `${this.baseUrl.replace(/\/$/, '')}/api/v1/connect/auth/token`,
        {
          client_id:     this.clientId,
          client_secret: this.clientSecret,
          grant_type:    'client_credentials',
        },
        { timeout: 15000, headers: { 'Content-Type': 'application/json' } }
      );

      const { access_token, expires_in = 3600 } = response.data;

      if (!access_token) {
        throw new AuthenticationError('Token response missing access_token field.');
      }

      this.cachedToken = access_token as string;
      this.expiresAt   = Math.floor(Date.now() / 1000) + (expires_in as number);

      return this.cachedToken;
    } catch (err: unknown) {
      if (axios.isAxiosError(err) && err.response) {
        const msg = (err.response.data as Record<string, unknown>)?.message as string ?? err.message;
        throw new AuthenticationError(`Failed to obtain access token: ${msg}`, err.response.status);
      }
      throw new AuthenticationError(`Failed to obtain access token: ${String(err)}`);
    }
  }

  revoke(): void {
    this.cachedToken = null;
    this.expiresAt   = 0;
  }
}

File: sdk/python/src/opescare/auth/token_manager.py
"""
OAuth2 client-credentials token manager.

Fetches a Bearer token from /api/v1/connect/auth/token,
caches it in memory, and refreshes 60 seconds before expiry.
"""
from __future__ import annotations

import time

import httpx

from ..exceptions import AuthenticationError


class TokenManager:
    _EXPIRY_BUFFER = 60  # refresh 60s before actual expiry

    def __init__(self, base_url: str, client_id: str, client_secret: str) -> None:
        self._base_url     = base_url.rstrip("/")
        self._client_id    = client_id
        self._client_secret = client_secret
        self._token: str | None = None
        self._expires_at: float = 0.0

    def get_token(self) -> str:
        """Return a valid token, refreshing if expired or near expiry."""
        if self._token and time.time() < (self._expires_at - self._EXPIRY_BUFFER):
            return self._token
        return self.refresh()

    def refresh(self) -> str:
        """Force a new token request."""
        url = f"{self._base_url}/api/v1/connect/auth/token"
        try:
            response = httpx.post(
                url,
                json={
                    "client_id":     self._client_id,
                    "client_secret": self._client_secret,
                    "grant_type":    "client_credentials",
                },
                headers={"Content-Type": "application/json"},
                timeout=15,
            )
            response.raise_for_status()
            data = response.json()
        except httpx.HTTPStatusError as exc:
            body    = _safe_json(exc.response)
            message = body.get("message", exc.response.text)
            raise AuthenticationError(
                f"Failed to obtain access token: {message}",
                exc.response.status_code,
                body,
            ) from exc
        except httpx.RequestError as exc:
            raise AuthenticationError(
                f"Failed to reach token endpoint: {exc}"
            ) from exc

        token = data.get("access_token")
        if not token:
            raise AuthenticationError(
                "Token response is missing 'access_token' field."
            )

        self._token      = token
        self._expires_at = time.time() + int(data.get("expires_in", 3600))
        return self._token

    def revoke(self) -> None:
        """Clear the cached token (forces a fresh fetch on next call)."""
        self._token      = None
        self._expires_at = 0.0


def _safe_json(response: httpx.Response) -> dict:
    try:
        return response.json()
    except Exception:
        return {}
